### ISSUE TITLE
Negative value in MOTA

### ISSUE THREAD

**Issue by mortezamg63** [cheind/py-motmetrics#47]
Hello
I am testing my tracker with groundtruth in training set, but the problem is that in some image sequences the MOTA is a negative value. Also, I have used SST tracker and its link is available from MOTChallenge website. The same result is shown for some training sequences. Could you guide me what is wrong here.

Thank you,
Morteza

**Comment by cheind** [comment]
Is this related to #18?

**Comment by mortezamg63** [comment]
Thanks for your response. The problem is odd. Your package and devkit in MOTChallenge dataset are working fine in MOT16 and MOT17 datasets. But, the problem of  negative values is in MOT15 dataset. I could not find a solution, but when I used MOT16 and MOT17 datasets, everything worked good.

Thank for your response

**Comment by cheind** [comment]
I think the problem is in missing preprocessing in py-motmetrics for MOT15. This preprocessing is always dataset specific and not the core development interest of py-motmetrics.

**Comment by naufil-hassan-confiz** [comment]
@cheind I'm also getting negative MOTA results on my own datasets. I have followed these steps to get the results

1.  (X,Y,width,height) format to calculated iou distance using "

> _mm.distances.iou_matrix_

" 
2.  populated my accumulator using
 "

> _acc.update( gt_labels_list, pr_labels_list, [ [dist from label-1 to all labels of prediction],  [dist from 
>   label-2 to all labels of prediction] ])_

"
3.  calculated MOT results using 

_

> mh = mm.metrics.create()
> summary = mh.compute_many(
>     [acc, acc.events], 
>     metrics=mm.metrics.motchallenge_metrics, 
>     generate_overall=True,
>     names=['full', 'part'])
> 
> strsummary = mm.io.render_summary(
>     summary, 
>     formatters=mh.formatters, 
>     namemap=mm.io.motchallenge_metric_names
> )
> print(strsummary)

_

4.  Results I get

         IDF1   IDP   IDR  Rcll  Prcn  GT MT  PT  ML   FP    FN IDs   FM  MOTA  MOTP
full    24.1% 29.8% 20.2% 32.1% 47.3% 176 11  56 109 3127  5928  94  258 -4.8% 0.341
part    24.1% 29.8% 20.2% 32.1% 47.3% 176 11  56 109 3127  5928  94  258 -4.8% 0.341
OVERALL 24.1% 29.8% 20.2% 32.1% 47.3% 352 22 112 218 6254 11856 188  516 -4.8% 0.341

**Comment by mortezamg63** [comment]
> I think the problem is in missing preprocessing in py-motmetrics for MOT15. This preprocessing is always dataset specific and not the core development interest of py-motmetrics.

Yes, I checked it and it seems the problem is preprocessing as you mentioned. Thanks for your reply

**Comment by eduardovelludo** [comment]
The MOTA metric is calculed based on misses, false positives and mismatches.
![mota](https://user-images.githubusercontent.com/6055460/68501149-393fd780-023c-11ea-9f60-0468d02bef61.gif)
Image from:
> Evaluating Multiple Object Tracking Performance: The CLEAR MOT Metrics

You can test your code's output by adding your misses, false positives and mismatches in the formula and verifying if the result is negative.

**Comment by EddieEduardo** [comment]
Unfortunately, I also got the negative values on MOTA/MOTAL when evaluating on my own dataset (I have already formatted .txt files into the same pattern with MOT16( gt.txt/det.txt/res.txt) )，can someone please tell me what's wrong with it ? Thanks a lot.
Here are my negative results:
**IDF1  IDP  IDR  | Rcll  Prcn  FAR   | GT   MT   PT   ML   | FP        FN       IDs   FM| MOTA  MOTP  MOTAL
19.7  25.0 16.3 | 20.1 30.9  4.96 |  11    2     1     8     | 2678   4744     3     65 | -25.0     73.3    -25.0**

**Comment by jvlmdr** [comment]
As @eduardovelludo pointed out, the range of MOTA is not [0, 1] but [-inf, 1].

```MOTA = 1 - (misses + false_positives + mismatches) / num_ground_truth```

It's entirely possible for the number of false positives to exceed the number of ground-truth boxes, and then you have MOTA < 0. For example, with 100 false positive detections and 10 true positives (with 0 misses and 0 mismatches) then `MOTA = 1 - 100 / 10 = -9`.

Note that it's even possible to MOTA = -inf when there are 0 ground-truth objects and 1 or more false positives.
